### PR TITLE
Mobile unit tests: remove custom waitFor implementation

### PR DESCRIPTION
**Pull request by jsnajdr**
The mobile tests, due to some past bugs in `@testing-library/react-native`, had their own implementation of `waitFor` and avoided using the `findBy*` query helpers, in favor of the less-recommended `waitFor(() => getBy*)` form.

This PR removes the custom `waitFor`. It was there because of bugs in `@testing-library/react-native` (some [fixed by myself](https://github.com/callstack/react-native-testing-library/releases/tag/v12.0.1) in 12.0.1 but there are others) that needed to be worked around. That's why this PR also upgrades the testing library to its latest version 12.1.2.

In addition to modifying tests, this PR also modifies the `useNestedSettingsUpdate` function. It used to retrieve the current value of `blockListSettings`, and to shallowly compare it with `newSettings`. But retrieving the `blockListSettings` with `useSelect` means that the component is rerendered after changing the settings, although it doesn't use them directly. This extra rerender was causing `act()` warnings. My patch is removing the retrieval and comparison, and just dispatches the `updateBlockListSettings` directly, even if the `newSettings` are the same. The state reducer itself will check, when handling `UPDATE_BLOCK_LIST_SETTINGS`, whether there are any real changes.

### CODE REVIEW

**Comment by jsnajdr** [comment]
One unit test is failing, auto-pasting of link URL in Embed block, and it's a bug in `@testing-library/react-native`, in their `waitFor` implementation. I'll submit a fix and will post a PR link here.

**Comment by jsnajdr** [comment]
Fixing the `@testing-library/react-native` bug here: https://github.com/callstack/react-native-testing-library/pull/1366

This PR is on hold until we can upgrade to a fixed version of the library.

**Comment by jsnajdr** [comment]
I had to change one more thing to make all tests green. Since v12.0.0, RNTL by default excludes elements hidden from accessibility from query results. And we need to find them explicitly with a `.findByText( 'foo' , { hidden: true } )` option.

Most of the fixes were in usages of the `RangeCell` component (quantities with units) which hides its internals from accessibility by adding a `importantForAccessibility="no-hide-descendants"` prop. Then it exposes special `accessibilityActions` for screen readers, like `activate` (to change units) or `increment`/`decrement` for changing the value. Note that there is no a11y action for editing the value. @AmandaRiu implemented this in #29741.

To test inaccessible parts of the UI, we need to declare explicitly we want to look at inaccessible elements. This doesn't mean that the UI wouldn't be accessible -- it just exposes two interfaces, accessible and "visual", and we want to test the visual one, too. The only shortcoming of `RangeCell` is that it doesn't support accessible editing of the value. Just changing the units and increment/decrement.

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (dcalhoun) [inline packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:119]: > In addition to modifying tests, this PR also modifies the `useNestedSettingsUpdate` function. It used to retrieve the current value of `blockListSettings`, and to shallowly compare it with `newSettings`. But retrieving the `blockListSettings` with `useSelect` means that the component is rerendered after changing the settings, although it doesn't use them directly. This extra rerender was causing `act()` warnings. My patch is removing the retrieval and comparison, and just dispatches the `updateBlockListSettings` directly, even if the `newSettings` are the same. The state reducer itself will check, when handling `UPDATE_BLOCK_LIST_SETTINGS`, whether there are any real changes.

Thank you for outlining the rationale behind this change above. 

For my own education, I posit the only "negative" of this change might be that we unnecessarily execute this loop when no settings have changed. I suppose the perception or reality is that the subsequent, unnecessary re-renders are more costly than the loop itself. Is that accurate?

File: packages/block-library/src/image/test/edit.native.js
Comment (dcalhoun) [inline packages/block-library/src/image/test/edit.native.js:205]: IIRC, this test's original intention was to verify the following flow: `None` → `Media File` → `Custom URL` → `Media File`. This was done because a scenario where not _all_ properties were correctly updated during the flow. 

Because of that, it may make most sense to reinstate `Media File` via two presses on that text. Once to select `Media File`, another to navigate back to the list of options. WDYT? 

<details><summary>Video of Flow</summary>

https://github.com/WordPress/gutenberg/assets/438664/bf8bd27f-1659-49ad-81cc-19c717780584

</details>

File: packages/block-library/src/list/test/edit.native.js
Comment (dcalhoun) [inline packages/block-library/src/list/test/edit.native.js:178]: This is a fantastic note and approach. Thank you for sharing this. 

@fluiddot it is likely beneficial to document this and leverage it in other scenarios.

File: packages/block-library/src/paragraph/test/edit.native.js
Comment (dcalhoun) [inline packages/block-library/src/paragraph/test/edit.native.js:254]: Noting this relates to https://github.com/WordPress/gutenberg/pull/51303#discussion_r1232058061, which refactors this test to interact with the element prior to it becoming inaccessible. I think it is fine for this PR to leave the `hidden` addition for now, but we should prefer avoiding its addition long term.

**Review by dcalhoun — APPROVED** [review]
This is fantastic! 💯 

I do not believe I can express how much this particular work means to me personally. This is one of those "thorns" that I have been unable to set aside time to address. **Thank you!** 🙇🏻 

I left a couple of comments, but I do not consider any of them blocking. So, with that, I approved these changes. I do not want to block merging or open up the potential for merge conflicts by unnecessarily postponing. 🚀

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (jsnajdr) [reply packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:119]: This change was guided not really by measurable performance concerns, but by the fact that the rerender triggered by a state update in a microtask leads to `act()` warnings. At the time when the rerender happens, there is no `act()` call active to catch it.

There are reasons to believe that an extra execution of the update loop is less costly that a React rerender. Because less code is executed. The worst thing that can happen is that they are both the same performance-wise.

File: packages/block-library/src/image/test/edit.native.js
Comment (jsnajdr) [reply packages/block-library/src/image/test/edit.native.js:205]: I'll have a second look at this test in the next few days.

I merged this PR now because I was afraid that it would get stalled again for a long time if I wanted to make everything perfect before merging.

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (dcalhoun) [reply packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:119]: Definitely understand the guidance or rationale. 

I suppose my point is that I found it interesting you modified the application code for a test. I generally avoid doing that. So, I likely would not have considered refactoring this code for a test, but instead leveraged additional `act` calls to resolve the warning. That is why I expressed what I perceived to be _potential_ negatives from the changes and sought your thoughts on those. 

With that clarification, this still seems like the best approach. 👍🏻

**Comment by tyxla** [comment]
So cool to see this land 👏

File: packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js
Comment (jsnajdr) [reply packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:119]: > I suppose my point is that I found it interesting you modified the application code for a test.

In this case I think the change is a good idea on its own, even without the affected test. The code used a pattern like this:
```js
const settings = useSelect( s => s( editor ).getSettings() );
const { setSetting } = useDispatch( editor );
function handleChange( key, value ) {
  if ( settings[ key ] !== value ) {
    setSetting( key, value );
  }
}
```
Here a React component is selecting the `settings` value, although it doesn't use it for render at all. Any change in `settings` will trigger a rerender that's guaranteed to be a noop. The value is used only to detect if the value is really changing.

Why not put the value change check directly into `setSetting`? The action and it's associated reducer has all the information it needs to do the check. And it avoids the rerender.
```js
const { setSetting } = useDispatch( editor );
function handleChange( key, value ) {
  setSetting( key, value );
}
```

Comment (dcalhoun) [reply packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:119]: That makes sense. Thank you for elaborating.

**Comment by fluiddot** [comment]
Just wanted to drop by to say thank you so much @jsnajdr for working on this, awesome job 🙇 . I'm glad we managed to remove the custom implementation of `waitFor` and place the native integration tests closer to the React Testing Library framework ❤️ .

**Comment by jsnajdr** [comment]
Thanks @fluiddot for your kind words :+1: If you ever need a helping hand with low level stuff like this, e.g., when upgrading React Native, please let me know.

**Pull request merged**
